boskos/ranch/ranch.go

```diff
@@ -106,24 +97,90 @@ func NewRanch(config string, storage string) (*Ranch, error) {
 //     owner - requester of the resource
 // Out: A valid Resource object on success, or
 //      ResourceNotFound error if target type resource does not exist in target state.
-func (r *Ranch) Acquire(rtype string, state string, dest string, owner string) (*common.Resource, error) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	for idx := range r.Resources {
-		res := r.Resources[idx]
-		if rtype == res.Type && state == res.State && res.Owner == "" {
-			res.LastUpdate = time.Now()
+func (r *Ranch) Acquire(rType, state, dest, owner string) (*common.Resource, error) {
+	r.resourcesLock.Lock()
+	defer r.resourcesLock.Unlock()
+
+	resources, err := r.Storage.GetResources()
+	if err != nil {
+		logrus.WithError(err).Errorf("could not get resources")
+		return nil, &ResourceNotFound{rType}
+	}
+
+	for idx := range resources {
+		res := resources[idx]
+		if rType == res.Type && state == res.State && res.Owner == "" {
+			res.LastUpdate = r.UpdateTime()
 			res.Owner = owner
 			res.State = dest
-
-			copy(r.Resources[idx:], r.Resources[idx+1:])
-			r.Resources[len(r.Resources)-1] = res
+			if err := r.Storage.UpdateResource(res); err != nil {
+				logrus.WithError(err).Errorf("could not update resource %s", res.Name)
+				return nil, err
+			}
 			return &res, nil
 		}
 	}
+	return nil, &ResourceNotFound{rType}
+}
+
+// AcquireByState checks out resources of a given typewithout an owner,
+// and move the checked out resource to the end of the resource list.
+// In: state - current state of the requested resource
+//     dest - destination state of the requested resource
+//     owner - requester of the resource
+//     names - names of resource to acquire
```

Comment:
so `names` is actually just a count for how many resources in this state you want to fetch?

Reply:
names is a list of resource names. So from Mason, we get the list from the leasedResources user data, and we should fetch the same with AcquireByState. But you could use this to get a subset.

Reply:
ack :-)